bootstrap/main.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2018 Open Source Foundries Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''West's bootstrap/wrapper script.
+'''
+
+import argparse
+import os
+from os.path import join
+import importlib
+import platform
+import subprocess
+import sys
+
+MANIFEST = 'manifest'
+MANIFEST_DEFAULT = 'https://github.com/zephyrproject-rtos/manifest'
+MANIFEST_REV_DEFAULT = 'master'
+WEST = 'west'
+WEST_DEFAULT = 'https://github.com/zephyrproject-rtos/west'
+WEST_REV_DEFAULT = 'master'
+if platform.system() == 'Windows':
+    # On Windows, the sentinel is hidden with attributes, rather than a
+    # leading '.', and thus doesn't contain one.
+    WEST_HIDDEN_SENTINEL = 'west-top'
+else:
+    # Assume a POSIX OS with leading '.' for hidden files.
+    WEST_HIDDEN_SENTINEL = '.west-top'
+
+#
+# Helpers shared between init and wrapper mode
+#
+
+
+class WestError(RuntimeError):
+    pass
+
+
+class WestNotFound(WestError):
+    '''Neither the current directory nor any parent has a West installation.'''
+
+
+def find_west_topdir(start):
+    '''Find the top-level installation directory, starting at ``start``.
+
+    If none is found, raises WestNotFound.'''
+    if not os.path.isdir(start):
+        raise WestNotFound()
+
+    if WEST_HIDDEN_SENTINEL in os.listdir(start):
+        return start
+
+    dirname = os.path.dirname(start)
+    if start == dirname:
+        # / on POSIX, top level drive name on Windows.
+        raise WestNotFound()
+    else:
+        return find_west_topdir(dirname)
+
+
+def make_topdir_sentinel(directory):
```

Comment:
Would be nice to have a description of why the sentinel is needed, when it's checked, etc. here.

Reply:
I thought it was obvious why by analogy to git and repo, but sure, I can add those.

Reply:
So it's like a `.git` for West, except still empty, and only used to flag a West installation?

Maybe it could be explained in those terms:

"foo/ is similar to a .git/ directory. Currently, it is only used to flag a directory as having a West installation in it. Files might be added to it in the future."

Could have `WEST_DIR` to take the analogy further maybe (similar to `GIT_DIR`). Not sure if that would get confusing though...